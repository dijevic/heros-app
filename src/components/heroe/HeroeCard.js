import React from 'react'
import { /* Link */ useHistory } from 'react-router-dom'
const heroesImages = require.context('../../assets/heroes', true)

export const HeroeCard = ({
    id,
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters }) => {
    const history = useHistory()
    const handleClick = ({ target }) => {

        // target.parentElement.firstElementChild.classList.add('outLeft')
        target.parentElement.classList.add('outLeft')

        setTimeout(() => {
            window.scrollTo(0, 0)
            history.push(`./hero/${id}`)
        }, 500)

        // elimine los link y agregue un div haciendo pruebas ***********NOTA************

    }

    return (
        <div /* to={} */ className="my-card" onClick={handleClick}>
            <img src={heroesImages(`./${id}.jpg`).default} className="img img-responsive" alt={superhero} loading="lazy" />
            <div className="profile-name">{superhero}</div>
            <div className="profile-position">{alter_ego}</div>
            <div className="profile-overview">
                <div className="profile-overview">
                    <div className="row">
                        <div className="col-ms-4">
                            <h3>{publisher}</h3>
                            <p>Primera aparici√≥n: <br />{first_appearance}</p>
                            {
                                (alter_ego !== characters)
                                && <p>{characters}</p>
                            }
                        </div>
                    </div>
                </div>
            </div>
        </div>
    )
}
