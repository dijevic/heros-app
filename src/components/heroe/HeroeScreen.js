import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router-dom'
import { getHeroeByid } from '../../selectors/GetHeroeById'

const heroesImages = require.context('../../assets/heroes', true)

export const HeroeScreen = ({ history }) => {
    console.log(heroesImages)
    const { id } = useParams()
    const heroe = useMemo(() => getHeroeByid(id), [id])


    if (!heroe || heroe.length === 0) {

        return <Redirect to="/" />
    }

    const { superhero, publisher, alter_ego, first_appearance, characters } = heroe[0]

    const handleClick = ({ target }) => {

        if (history.length <= 2) {
            history.push('/')
        } else {
            target.parentElement.parentElement.firstElementChild.classList.add('outLeft')
            target.parentElement.parentElement.firstElementChild.nextElementSibling.classList.add('outDown')

            setTimeout(() => {
                history.goBack()

            }, 500);
        }


    }
    return (
        <div className="row mt-5 heroe-container">

            <div className="col-4">
                {/* <img src={`../assets/heroes/${id}.jpg`} alt={superhero} className="img-thumbnail " /> */}
                <img src={heroesImages(`./${id}.jpg`).default} alt={superhero} className="img-thumbnail " />
            </div>

            <div className="col-8">
                <h3>{superhero}</h3>

                <ul className="list-group list-group-flush">
                    <li className="list-group-item"> <b>Alter ego:</b> {alter_ego}</li>
                    <li className="list-group-item"><b>publisher:</b> {publisher}</li>
                    <li className="list-group-item"><b>first appearance:</b>{first_appearance}</li>
                </ul>

                <h5>characters</h5>
                <p>{characters}</p>

                <button className="btn btn-primary" onClick={handleClick}>return</button>
            </div>

        </div>
    )
}
