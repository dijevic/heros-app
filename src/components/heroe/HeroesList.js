import React, { useMemo } from 'react'
import { getHeroeByPublisher } from '../../selectors/GetHeroeByPublisher'
import { HeroeCard } from './HeroeCard'

export const HeroesList = ({ publisher, history }) => {


    const heroes = useMemo(() => getHeroeByPublisher(publisher), [publisher])


    return (
        <div className="card-columns row row-cols-1 row-cols-md-3 g-4">
            {
                heroes.map(hero => {
                    return <HeroeCard
                        key={hero.id}
                        {...hero}

                    >{hero.superhero}</HeroeCard>

                })

            }
        </div>
    )
}
