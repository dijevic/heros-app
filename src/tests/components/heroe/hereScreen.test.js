import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { HeroeScreen } from "../../../components/heroe/HeroeScreen"

describe('pruebas en mi componente de hero screen', () => {




    test('debe de mostrar mi componente redirect', () => {
        const historyMock = {
            push: jest.fn(),
            length: 10,
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroeScreen history={historyMock} />
            </MemoryRouter>
        )
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('Redirect').exists()).toBe(true)
    })

    test('debe de renderizar un heroe y aceptar parametros por el route', () => {
        const historyMock = {
            push: jest.fn(),
            length: 10,
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/dc-batman']}>
                <Route path='/hero/:id'>
                    <HeroeScreen history={historyMock} />
                </Route>
            </MemoryRouter>
        )


        expect(wrapper.find('.row').exists()).toBe(true)

    })

    test('debe de llamar al metodo push de mi history', () => {
        const historyMock = {
            push: jest.fn(),
            length: 1,
            goBack: jest.fn()
        }


        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/dc-batman']}>
                <Route path='/hero/:id'>
                    <HeroeScreen history={historyMock} />
                </Route>
            </MemoryRouter>
        )

        wrapper.find('button').simulate('click')
        expect(historyMock.push).toHaveBeenCalled()
        expect(historyMock.goBack).not.toHaveBeenCalled()


    })
    test('debe de llamar al metodo goBack de mi history', () => {
        const historyMock = {
            push: jest.fn(),
            length: 10,
            goBack: jest.fn()
        }


        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/dc-batman']}>
                <Route path='/hero/:id'>
                    <HeroeScreen history={historyMock} />
                </Route>
            </MemoryRouter>
        )

        wrapper.find('button').simulate('click')
        expect(historyMock.goBack).toHaveBeenCalled()

        expect(historyMock.push).not.toHaveBeenCalled()


    })

    test('debe de mostrar mi componente redirect si el heroe no existe', () => {
        const historyMock = {
            push: jest.fn(),
            length: 10,
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/32132123123']}>
                <HeroeScreen history={historyMock} />
            </MemoryRouter>
        )
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('Redirect').exists()).toBe(true)
    })
})