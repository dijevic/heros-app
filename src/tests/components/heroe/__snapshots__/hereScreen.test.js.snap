// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en mi componente de hero screen debe de mostrar mi componente redirect 1`] = `""`;

exports[`pruebas en mi componente de hero screen debe de mostrar mi componente redirect si el heroe no existe 1`] = `""`;
